runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/io/GzipSdkSource.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.InternalApi
+import java.util.zip.GZIPOutputStream
+
+@InternalApi
+public class GzipSdkSource(
+    private val source: SdkSource,
+    private val bytesAvailable: Long?,
+) : SdkSource {
+    private val gzipBuffer = SdkBuffer()
+    private val gzipOutputStream = GZIPOutputStream(gzipBuffer.outputStream(), true)
+    private var bytesRead: Long = 0
+
+    /**
+     * Keeps track of whether a read operation has been made on this sdk source
+     */
+    private var read: Boolean = false
+
+    override fun read(sink: SdkBuffer, limit: Long): Long {
+        require(limit >= 0L)
+        if (limit == 0L) return 0L
+
+        if (bytesRead == bytesAvailable) {
+            if (!read) { // Empty payload
+                gzipOutputStream.write(ByteArray(0))
+                gzipOutputStream.close()
+                gzipBuffer.readAll(sink)
+                gzipBuffer.close()
+
+                read = true
+            }
+
+            return -1
+        }
+
+        if (!read) read = true
+
+        val temp = SdkBuffer()
+        val rc = source.read(temp, limit)
+
+        gzipOutputStream.write(temp.readByteArray())
+        gzipBuffer.readAll(sink)
+
+        bytesRead += rc
+
+        if (bytesRead == bytesAvailable || rc == -1L) {
```

Comment:
question/correctness: What happens when `bytesAvailable` is not provided (null)? Does this still work? Maybe it could be worth adding to your test suite

Reply:
Yes, I think this still works, it would use the `rc` value to check for -1 and determine the source is read completely that way. I was testing it without the `bytesRead` and `bytesAvailable` logic before and it worked.  You're right about adding a test for that though, I think it should be in the test suite. 

The only reason that additional logic (`bytesAvailable` & `bytesRead`) is there is because when using `readFully(...)` when the source is read completely there is no -1 returned. The -1 is returned only after exactly all the bytes in the source are read and then another read operation is tried. `readFully()` considers -1 as an error, so the additional logic catches when we read the source completely and does all the closing and finalizing. 

The only case where this wouldn't work is when there is a chunked payload which has unknown length and someone would try to read the whole payload using `readFully()`. Then we would never know when to finalize and close the gzip output stream and buffer because a -1 throw an exception before we could do that.